### PR TITLE
Documentation update; Fix connect_fs_neat, full, partial so follow documentation, while retaining compatibility

### PR DESCRIPTION
**Pull request by drallensmith**
(Re `full` and `partial` - see my next set of comments.)
`connect_fs_neat` was only connecting the selected input node to all output nodes, not to both hidden and output nodes. I have put in two alternate selections - fs_neat_nohidden (as per how connect_fs_neat worked) and fs_neat_hidden (as per the documentation on fs_neat); the latter does connect to hidden nodes. (Without hidden nodes, they are identical.) If someone's configuration asks for both fs_neat and hidden nodes, they will now get a warning telling them to use either fs_neat_nohidden or fs_neat_hidden. I will update the documentation for this along with the other changes in the documentation that I'm working on. (After I'm finished with the documentation, I may try my hand at writing tests for fs_neat in test_genome.py.)

I have also made the get_new_node_key revision more efficient, and added a bit of info to the docstring for compute_full_connections (as long as I was revising code in the vicinity...).

-Allen

P.S. You were mentioning the QuantifiedCode (or whatever) as being to help remind you to keep things more documented? Heh. I'd be scared of overloading pydoc if we put **all** the documentation (even with sphinx-specific code removed) I've written into the modules... some of it should be usable for that, however (and some of it is from docstrings that are already there).

### CODE REVIEW

**Comment by drallensmith** [comment]
As it turns out, there was a related (original) documentation vs code issue for the full (and by extension partial) initial_connectivity setting - whether there were (with hidden nodes present) direct connections between input and output nodes.

-Allen

**Comment by drallensmith** [comment]
Oh. For these, I did update the `tests/test_genome.py` file.

-Allen

**Comment by drallensmith** [comment]
I should probably have done `compute_full_connections_nodirect` and `compute_full_connections_direct` as one `compute_full_connections(self, config, direct)` method, with `direct` as a boolean. Want me to change it?

-Allen

**Comment by drallensmith** [comment]
I goofed in the warning output for 'partial' in genome.py - sorry! Fortunately, I got compulsive and decided to add in tests for the old cases (it would be nice to test for whether the warning got emitted, of course...); I also added tests for fs_neat, and did the merger I mentioned above. (I'll merge the other _hidden/_nohidden and _direct/_nodirect if desired.)

-Allen

**Comment by drallensmith** [comment]
Various pylint fixes - docstrings and similar. I refrained from putting in one commit that would conflict with the PR @bennr01 put in (will have to investigate if can do commit of all but one file...)

**Comment by bennr01** [comment]
@drallensmith are you using the `git` command or some other software? IIRC you can use `git add <filename>` to add only the specified file to the commit.
I also found [this Stackoverflow Question](https://stackoverflow.com/questions/12481639/remove-files-from-git-commit/15321456#15321456) about removing a file from a commit.
Also, *sometimes* it may help to merge the conflicting branch into the current branch.

**Comment by drallensmith** [comment]
Sorry, I wasn't clear. The commits in questions are ones I've already done to other branches (docs_update and config_work) in my local and github repositories - I simply added them to the master branch to help reduce the number of PRs @CodeReclaimers needs to deal with. I'll take a look at the StackOverflow post - thanks!

**Comment by drallensmith** [comment]
As the commits state, the additions are:

- Adding docstrings to test_genome.py, since the additions were making it get complicated
- Changes to picture2d accounting for the partial docs vs code mixup corrected by the rest of the patches; in my admittedly-biased opinion, the results are also better (more interaction between the dimensions, for instance).

**Comment by drallensmith** [comment]
My fork's docs_update branch appears to finally be stable enough to be brought in. As mentioned in #92 you can take a look at the formatted results (with the exception of one note re my fork's config_work branch) on my github.io page. It is mainly changes in the docs directory, as one might guess, but also adds some additional docstrings. I don't **_think_** it will conflict with the parallel.py changes @bennr01 has done, but I could be wrong - sorry!

**Comment by bennr01** [comment]
@drallensmith You dont have to worry about possible merge conflicts with my changes on `parallel.py`. If my PR would result in a merge conflict, i would just try to fix it myself. If everyone would have to check that their changes would not conflict with the changes of any other contributor, large projects would be nearly impossible to maintain. And your english is probably way better than mine, so your docstring changes would probably be better than the ones i made. But thanks for worrying :smile:.

**Comment by drallensmith** [comment]
True re projects; good point, thanks! My instinct is to try to avoid giving other people extra work, but I can see where it's not always practical. Your English is fine, from what I've seen - mine tends to be too wordy, actually. (This even extends to programming - my variable names tend to be rather long, if - as intended - clear in what they're doing, at least to me!) And quite welcome re the consideration/worrying.

**Comment by CodeReclaimers** [comment]
Thanks again for all the work on this--since I haven't had time to look, I figured I might as well just merge and let people try it out. :)
